File: desktop/src/features/sidebar/lib/channelSectionsStorage.ts
import { normalizeRelayUrl } from "@/shared/lib/normalizeRelayUrl";

const STORAGE_KEY_PREFIX = "buzz-channel-sections.v1";
export const MAX_CHANNEL_SECTIONS = 100;
export const MAX_CHANNEL_SECTION_ASSIGNMENTS = 1_000;
export const MAX_MANUALLY_UNASSIGNED_CHANNELS = 1_000;
export const MAX_SMART_SECTION_PATTERN_LENGTH = 200;

export type SmartSectionRule = {
  pattern: string;
  isRegex: boolean;
  caseSensitive: boolean;
};

export type ChannelSection = {
  id: string;
  name: string;
  icon?: string;
  order: number;
  smartRule?: SmartSectionRule;
};

export type ChannelSectionStore = {
  version: 1;
  sections: ChannelSection[];
  assignments: Record<string, string>;
  /** Channels explicitly moved to the default Channels group by the user. */
  manuallyUnassignedChannelIds?: string[];
};

export const DEFAULT_STORE: ChannelSectionStore = Object.freeze({
  version: 1,
  sections: [],
  assignments: {},
});

/**
 * Returns the localStorage key for channel sections.
 *
 * When `relayUrl` is provided the key is scoped to that relay (normalized via
 * the same `normalizeRelayUrl` used by all relay-scoped local stores) so
 * sections from different communities/relays don't bleed across each other.
 * When omitted the legacy pubkey-only key is returned (used only during
 * one-time migration in `readChannelSectionsStore`).
 */
export function storageKey(pubkey: string, relayUrl?: string): string {
  if (!relayUrl) return `${STORAGE_KEY_PREFIX}:${pubkey}`;
  const normalized = normalizeRelayUrl(relayUrl);
  // Encode the normalized relay so it can't contain the `:` delimiter.
  return `${STORAGE_KEY_PREFIX}:${pubkey}:${encodeURIComponent(normalized)}`;
}

export function boundChannelSectionsStore(
  store: ChannelSectionStore,
): ChannelSectionStore {
  const sections = store.sections
    .slice()
    .sort((left, right) => left.order - right.order)
    .slice(-MAX_CHANNEL_SECTIONS);
  const sectionIds = new Set(sections.map((section) => section.id));
  const assignments = Object.fromEntries(
    Object.entries(store.assignments)
      .filter(([, sectionId]) => sectionIds.has(sectionId))
      .slice(-MAX_CHANNEL_SECTION_ASSIGNMENTS),
  );
  const manuallyUnassignedChannelIds = Array.from(
    new Set(store.manuallyUnassignedChannelIds ?? []),
  ).slice(-MAX_MANUALLY_UNASSIGNED_CHANNELS);
  if (
    sections.length === store.sections.length &&
    Object.keys(assignments).length === Object.keys(store.assignments).length &&
    manuallyUnassignedChannelIds.length ===
      (store.manuallyUnassignedChannelIds?.length ?? 0)
  ) {
    return store;
  }
  return {
    ...store,
    sections,
    assignments,
    ...(manuallyUnassignedChannelIds.length > 0
      ? { manuallyUnassignedChannelIds }
      : { manuallyUnassignedChannelIds: undefined }),
  };
}

export function stripOrphanedAssignments(
  store: ChannelSectionStore,
): ChannelSectionStore {
  const sectionIds = new Set(store.sections.map((s) => s.id));
  const cleaned = Object.fromEntries(
    Object.entries(store.assignments).filter(([, sid]) => sectionIds.has(sid)),
  );
  const stripped =
    Object.keys(cleaned).length === Object.keys(store.assignments).length
      ? store
      : { ...store, assignments: cleaned };
  return boundChannelSectionsStore(stripped);
}

export function parseChannelSectionPayload(
  json: unknown,
): ChannelSectionStore | null {
  if (typeof json !== "object" || json === null) return null;
  const obj = json as Record<string, unknown>;
  const sections: ChannelSection[] = Array.isArray(obj.sections)
    ? obj.sections.flatMap((entry: unknown): ChannelSection[] => {
        if (typeof entry !== "object" || entry === null) return [];
        const section = entry as Record<string, unknown>;
        if (
          typeof section.id !== "string" ||
          typeof section.name !== "string" ||
          typeof section.order !== "number"
        ) {
          return [];
        }
        const icon =
          typeof section.icon === "string" && section.icon.trim().length > 0
            ? section.icon.trim()
            : undefined;
        const smartRule = parseSmartSectionRule(section.smartRule);
        return [
          {
            id: section.id,
            name: section.name,
            ...(icon ? { icon } : {}),
            order: section.order,
            ...(smartRule ? { smartRule } : {}),
          },
        ];
      })
    : [];
  const assignments: Record<string, string> =
    typeof obj.assignments === "object" &&
    obj.assignments !== null &&
    !Array.isArray(obj.assignments)
      ? Object.fromEntries(
          Object.entries(obj.assignments as Record<string, unknown>).filter(
            (entry): entry is [string, string] => typeof entry[1] === "string",
          ),
        )
      : {};
  const manuallyUnassignedChannelIds = Array.isArray(
    obj.manuallyUnassignedChannelIds,
  )
    ? Array.from(
        new Set(
          obj.manuallyUnassignedChannelIds.filter(
            (channelId): channelId is string => typeof channelId === "string",
          ),
        ),
      ).slice(-MAX_MANUALLY_UNASSIGNED_CHANNELS)
    : [];
  return stripOrphanedAssignments({
    version: 1,
    sections,
    assignments,
    ...(manuallyUnassignedChannelIds.length > 0
      ? { manuallyUnassignedChannelIds }
      : {}),
  });
}

function parseSmartSectionRule(value: unknown): SmartSectionRule | undefined {
  if (typeof value !== "object" || value === null) return undefined;
  const rule = value as Record<string, unknown>;
  if (
    typeof rule.pattern !== "string" ||
    rule.pattern.length === 0 ||
    rule.pattern.length > MAX_SMART_SECTION_PATTERN_LENGTH ||
    typeof rule.isRegex !== "boolean" ||
    typeof rule.caseSensitive !== "boolean"
  ) {
    return undefined;
  }
  return {
    pattern: rule.pattern,
    isRegex: rule.isRegex,
    caseSensitive: rule.caseSensitive,
  };
}

function parseRaw(raw: string | null): ChannelSectionStore | null {
  if (!raw) return null;
  try {
    const parsed = JSON.parse(raw);
    if (typeof parsed !== "object" || parsed === null || parsed.version !== 1) {
      return null;
    }
    return parseChannelSectionPayload(parsed);
  } catch {
    return null;
  }
}

/**
 * Read the section store for `pubkey` scoped to `relayUrl`.
 *
 * On first access for a scoped key, migrates any existing data from the
 * legacy pubkey-only key so users don't lose their sections on upgrade.
 * After a successful migration write the legacy key is deleted, making
 * this a globally one-time operation — subsequent empty scoped-key reads
 * (i.e. a different relay) won't see legacy data and won't trigger
 * cross-relay contamination via the seed-publish path.
 */
export function readChannelSectionsStore(
  pubkey: string,
  relayUrl?: string,
): ChannelSectionStore {
  try {
    const key = storageKey(pubkey, relayUrl);
    const raw = window.localStorage.getItem(key);

    // Scoped key already has data — use it directly.
    if (raw !== null) {
      return parseRaw(raw) ?? DEFAULT_STORE;
    }

    // No scoped data yet.  If we were given a relay scope, attempt a
    // one-time migration from the legacy pubkey-only key.
    if (relayUrl) {
      const legacyKey = storageKey(pubkey);
      const legacyRaw = window.localStorage.getItem(legacyKey);
      const migrated = parseRaw(legacyRaw);
      if (migrated && migrated.sections.length > 0) {
        // Persist under the scoped key and remove the legacy key so this
        // migration cannot fire again for any other relay scope.
        try {
          window.localStorage.setItem(key, JSON.stringify(migrated));
          window.localStorage.removeItem(legacyKey);
        } catch {
          // Ignore write failures — we still return the migrated value.
        }
        return migrated;
      }
    }

    return DEFAULT_STORE;
  } catch {
    return DEFAULT_STORE;
  }
}

export function writeChannelSectionsStore(
  pubkey: string,
  store: ChannelSectionStore,
  relayUrl?: string,
): boolean {
  try {
    window.localStorage.setItem(
      storageKey(pubkey, relayUrl),
      JSON.stringify(boundChannelSectionsStore(store)),
    );
    return true;
  } catch {
    return false;
  }
}

File: mobile/lib/features/channels/channel_sections/channel_sections_storage.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

/// Normalizes a relay URL for storage keys so equivalent URLs share a scope.
/// Matches desktop `normalizeRelayUrl` and mobile channel-sort.
String normalizeChannelSectionsRelayUrl(String relayUrl) =>
    relayUrl.trim().replaceFirst(RegExp(r'/+$'), '').toLowerCase();

/// Relay-scoped local storage key for channel sections.
String channelSectionsKey(String pubkey, String relayUrl) =>
    'buzz.channel-sections.v1:$pubkey:${Uri.encodeComponent(normalizeChannelSectionsRelayUrl(relayUrl))}';

/// Pre-scoping key (account-global). Used only for one-time migration.
String legacyChannelSectionsKey(String pubkey) =>
    'buzz.channel-sections.v1:$pubkey';

class ChannelSection {
  final String id;
  final String name;
  final String? icon;
  final int order;

  const ChannelSection({
    required this.id,
    required this.name,
    this.icon,
    required this.order,
  });

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    if (icon != null) 'icon': icon,
    'order': order,
  };

  factory ChannelSection.fromJson(Map<String, dynamic> json) => ChannelSection(
    id: json['id'] as String,
    name: json['name'] as String,
    icon: json['icon'] is String ? json['icon'] as String : null,
    order: json['order'] as int,
  );
}

class ChannelSectionStore {
  final int version;
  final List<ChannelSection> sections;
  final Map<String, String> assignments;

  const ChannelSectionStore({
    this.version = 1,
    this.sections = const [],
    this.assignments = const {},
  });

  Map<String, dynamic> toJson() => {
    'version': version,
    'sections': sections.map((s) => s.toJson()).toList(),
    'assignments': assignments,
  };

  factory ChannelSectionStore.fromJson(Map<String, dynamic> json) {
    final rawSections = json['sections'];
    final sections = <ChannelSection>[];
    if (rawSections is List) {
      for (final entry in rawSections) {
        if (entry is Map<String, dynamic> &&
            entry['id'] is String &&
            entry['name'] is String &&
            entry['order'] is int) {
          sections.add(ChannelSection.fromJson(entry));
        }
      }
    }

    final rawAssignments = json['assignments'];
    final assignments = <String, String>{};
    if (rawAssignments is Map) {
      for (final entry in rawAssignments.entries) {
        if (entry.key is String && entry.value is String) {
          assignments[entry.key as String] = entry.value as String;
        }
      }
    }

    // Strip assignments referencing sections that don't exist.
    final sectionIds = {for (final s in sections) s.id};
    assignments.removeWhere((_, sectionId) => !sectionIds.contains(sectionId));

    return ChannelSectionStore(
      version: 1,
      sections: sections,
      assignments: assignments,
    );
  }
}

class ChannelSectionsStorage {
  final SharedPreferences _prefs;

  ChannelSectionsStorage(this._prefs);

  /// Read sections for [pubkey] scoped to [relayUrl].
  ///
  /// On first access for a scoped key, migrates any existing data from the
  /// legacy pubkey-only key so users don't lose their sections on upgrade.
  /// After a successful migration the legacy key is deleted — subsequent
  /// empty scoped-key reads (a different relay) won't see legacy data.
  ChannelSectionStore read(String pubkey, String relayUrl) {
    final scoped = _readKey(channelSectionsKey(pubkey, relayUrl));
    if (scoped != null) return scoped;

    // One-time migration from account-global storage (pre-relay-scope).
    // The first active relay claims the legacy value; removing it prevents
    // the same unscoped sections from bleeding into later communities.
    final legacy = _readKey(legacyChannelSectionsKey(pubkey));
    if (legacy != null && legacy.sections.isNotEmpty) {
      write(pubkey, relayUrl, legacy);
      _prefs.remove(legacyChannelSectionsKey(pubkey));
      return legacy;
    }

    return const ChannelSectionStore();
  }

  ChannelSectionStore? _readKey(String key) {
    final raw = _prefs.getString(key);
    if (raw == null || raw.isEmpty) {
      return null;
    }

    try {
      final parsed = jsonDecode(raw);
      if (parsed is! Map<String, dynamic>) {
        return null;
      }
      if (parsed['version'] != 1) {
        return null;
      }
      return ChannelSectionStore.fromJson(parsed);
    } catch (_) {
      return null;
    }
  }

  void write(String pubkey, String relayUrl, ChannelSectionStore store) {
    _prefs.setString(
      channelSectionsKey(pubkey, relayUrl),
      jsonEncode(store.toJson()),
    );
  }
}
